unikernel.ml

```diff
@@ -490,39 +497,186 @@ struct
+                    in
+                    let validate_user_input email password =
+                      if email = "" || password = "" then
+                        Error "All fields must be filled."
+                      else if not (Re.execp validate_email_re email) then
+                        Error "Invalid email address."
+                      else if String.length password < 8 then
+                        Error "Password must be at least 8 characters long."
+                      else Ok "Validation passed."
+                    in
+                    let email =
+                      json
+                      |> Yojson.Basic.Util.member "email"
+                      |> Yojson.Basic.to_string
+                    in
+                    let password =
+                      json
+                      |> Yojson.Basic.Util.member "password"
+                      |> Yojson.Basic.to_string
+                    in
+                    let validate_user = validate_user_input email password in
+                    match validate_user with
+                    | Error s ->
+                        let res =
+                          "{\"status\": 400, \"success\": false, \"message\": \
+                           \"" ^ s ^ "\"}"
+                        in
+                        Lwt.return (reply ~content_type:"application/json" res)
+                    | Ok _ -> (
+                        let _, (t : Storage.t) = !store in
+                        let users = t.users in
+                        let login =
+                          User_model.login_user ~email ~password users
+                        in
+                        match login with
+                        | Error (`Msg s) ->
                             let res =
-                              "{\"status\": 400, \"success\": false, \
-                               \"message\": \"Something went wrong. Wait a few \
-                               seconds and try again.\"}"
+                              "{\"status\": 404, \"message\": \"" ^ s ^ "\"}"
                             in
                             Lwt.return
-                              (reply ~content_type:"application/json" res))))
+                              (reply ~content_type:"application/json" res)
+                        | Ok user -> (
+                            Store.update_user !store user >>= function
+                            | Ok store' ->
+                                store := store';
+                                let res =
+                                  "{\"status\": 200, \"success\": true, \
+                                   \"message\": {\"user\": "
+                                  ^ Yojson.Basic.to_string
+                                      (User_model.user_to_json user)
+                                  ^ "}}"
+                                in
+                                let cookie =
+                                  List.find
```

Comment:
with this, we're very sure that such a cookie exists? I'd suggest to use a List.find_opt, and later revise the data structure to always have such a field :)